File: apps/backend-api/internal/service/pedigree.go
package service

import (
	"context"

	"github.com/google/uuid"
	"github.com/schema-platform/backend-api/internal/dto"
	"github.com/schema-platform/backend-api/internal/model"
	"github.com/schema-platform/backend-api/internal/repository"
	"github.com/schema-platform/backend-api/pkg/errors"
)

// PedigreeService handles pedigree operations
type PedigreeService struct {
	pedigreeRepo *repository.PedigreeRepository
	memberRepo   *repository.PedigreeMemberRepository
	sampleRepo   *repository.SampleRepository
}

// NewPedigreeService creates a new pedigree service
func NewPedigreeService(pedigreeRepo *repository.PedigreeRepository, memberRepo *repository.PedigreeMemberRepository, sampleRepo *repository.SampleRepository) *PedigreeService {
	return &PedigreeService{
		pedigreeRepo: pedigreeRepo,
		memberRepo:   memberRepo,
		sampleRepo:   sampleRepo,
	}
}

// CreatePedigree creates a new pedigree
func (s *PedigreeService) CreatePedigree(ctx context.Context, req *dto.PedigreeCreateRequest) (*dto.PedigreeResponse, error) {
	// Check if name exists
	exists, err := s.pedigreeRepo.ExistsByName(ctx, req.Name)
	if err != nil {
		return nil, errors.WrapDatabaseError(err)
	}
	if exists {
		return nil, errors.NewConflictError("Pedigree name already exists")
	}

	pedigree := &model.Pedigree{
		Name:    req.Name,
		Disease: req.Disease,
		Note:    req.Note,
	}

	if err := s.pedigreeRepo.Create(ctx, pedigree); err != nil {
		return nil, errors.WrapDatabaseError(err)
	}

	return s.toPedigreeResponse(pedigree), nil
}

// GetPedigree retrieves a pedigree by ID
func (s *PedigreeService) GetPedigree(ctx context.Context, id string) (*dto.PedigreeDetailResponse, error) {
	pedigreeID, err := uuid.Parse(id)
	if err != nil {
		return nil, errors.NewValidationError("Invalid pedigree ID")
	}

	pedigree, err := s.pedigreeRepo.GetByID(ctx, pedigreeID)
	if err != nil {
		return nil, errors.WrapDatabaseError(err)
	}

	return s.toPedigreeDetailResponse(pedigree), nil
}

// ListPedigrees lists all pedigrees
func (s *PedigreeService) ListPedigrees(ctx context.Context, page, pageSize int) (*dto.PaginatedResponse[dto.PedigreeResponse], error) {
	offset := (page - 1) * pageSize
	pedigrees, total, err := s.pedigreeRepo.List(ctx, pageSize, offset)
	if err != nil {
		return nil, errors.WrapDatabaseError(err)
	}

	items := make([]dto.PedigreeResponse, len(pedigrees))
	for i, p := range pedigrees {
		items[i] = *s.toPedigreeResponse(&p)
	}

	return dto.NewPaginatedResponse(items, total, page, pageSize), nil
}

// UpdatePedigree updates a pedigree
func (s *PedigreeService) UpdatePedigree(ctx context.Context, id string, req *dto.PedigreeUpdateRequest) (*dto.PedigreeResponse, error) {
	pedigreeID, err := uuid.Parse(id)
	if err != nil {
		return nil, errors.NewValidationError("Invalid pedigree ID")
	}

	pedigree, err := s.pedigreeRepo.GetByID(ctx, pedigreeID)
	if err != nil {
		return nil, errors.WrapDatabaseError(err)
	}

	if req.Name != nil && *req.Name != pedigree.Name {
		exists, err := s.pedigreeRepo.ExistsByName(ctx, *req.Name)
		if err != nil {
			return nil, errors.WrapDatabaseError(err)
		}
		if exists {
			return nil, errors.NewConflictError("Pedigree name already exists")
		}
		pedigree.Name = *req.Name
	}

	if req.Disease != nil {
		pedigree.Disease = *req.Disease
	}
	if req.Note != nil {
		pedigree.Note = *req.Note
	}

	if err := s.pedigreeRepo.Update(ctx, pedigree); err != nil {
		return nil, errors.WrapDatabaseError(err)
	}

	return s.toPedigreeResponse(pedigree), nil
}

// DeletePedigree deletes a pedigree
func (s *PedigreeService) DeletePedigree(ctx context.Context, id string) error {
	pedigreeID, err := uuid.Parse(id)
	if err != nil {
		return errors.NewValidationError("Invalid pedigree ID")
	}

	return s.pedigreeRepo.Delete(ctx, pedigreeID)
}

// AddMember adds a member to a pedigree
func (s *PedigreeService) AddMember(ctx context.Context, pedigreeID string, req *dto.PedigreeMemberCreateRequest) (*dto.PedigreeMemberResponse, error) {
	parsedID, err := uuid.Parse(pedigreeID)
	if err != nil {
		return nil, errors.NewValidationError("Invalid pedigree ID")
	}

	pedigree, err := s.pedigreeRepo.GetByID(ctx, parsedID)
	if err != nil {
		return nil, errors.WrapDatabaseError(err)
	}

	member := &model.PedigreeMember{
		PedigreeID:     pedigree.ID,
		Name:           req.Name,
		Gender:         req.Gender,
		BirthYear:      req.BirthYear,
		IsDeceased:     req.IsDeceased,
		DeceasedYear:   req.DeceasedYear,
		Relation:       model.RelationType(req.Relation),
		AffectedStatus: model.AffectedStatus(req.AffectedStatus),
		Phenotypes:     req.Phenotypes,
		FatherID:       req.FatherID,
		MotherID:       req.MotherID,
		Generation:     req.Generation,
		Position:       req.Position,
	}

	if req.SampleID != nil {
		sampleUUID, err := uuid.Parse(*req.SampleID)
		if err != nil {
			return nil, errors.NewValidationError("Invalid sample ID")
		}
		member.SampleID = &sampleUUID
	}

	if err := s.memberRepo.Create(ctx, member); err != nil {
		return nil, errors.WrapDatabaseError(err)
	}

	// If this is the proband, update the pedigree
	if req.Relation == "proband" {
		pedigree.ProbandMemberID = &member.ID
		if err := s.pedigreeRepo.Update(ctx, pedigree); err != nil {
			return nil, errors.WrapDatabaseError(err)
		}
	}

	return s.toMemberResponse(member), nil
}

// GetMembers retrieves all members of a pedigree
func (s *PedigreeService) GetMembers(ctx context.Context, pedigreeID string) ([]dto.PedigreeMemberResponse, error) {
	parsedID, err := uuid.Parse(pedigreeID)
	if err != nil {
		return nil, errors.NewValidationError("Invalid pedigree ID")
	}

	members, err := s.memberRepo.GetByPedigreeID(ctx, parsedID)
	if err != nil {
		return nil, errors.WrapDatabaseError(err)
	}

	items := make([]dto.PedigreeMemberResponse, len(members))
	for i, m := range members {
		items[i] = *s.toMemberResponse(&m)
	}

	return items, nil
}

// UpdateMember updates a pedigree member
func (s *PedigreeService) UpdateMember(ctx context.Context, memberID string, req *dto.PedigreeMemberUpdateRequest) (*dto.PedigreeMemberResponse, error) {
	memberIDParsed, err := uuid.Parse(memberID)
	if err != nil {
		return nil, errors.NewValidationError("Invalid member ID")
	}

	member, err := s.memberRepo.GetByID(ctx, memberIDParsed)
	if err != nil {
		return nil, errors.WrapDatabaseError(err)
	}

	// Update fields
	if req.Name != nil {
		member.Name = *req.Name
	}
	if req.Gender != nil {
		member.Gender = *req.Gender
	}
	if req.BirthYear != nil {
		member.BirthYear = req.BirthYear
	}
	if req.IsDeceased != nil {
		member.IsDeceased = *req.IsDeceased
	}
	if req.DeceasedYear != nil {
		member.DeceasedYear = req.DeceasedYear
	}
	if req.Relation != nil {
		member.Relation = model.RelationType(*req.Relation)
	}
	if req.AffectedStatus != nil {
		member.AffectedStatus = model.AffectedStatus(*req.AffectedStatus)
	}
	if req.Phenotypes != nil {
		member.Phenotypes = *req.Phenotypes
	}
	if req.FatherID != nil {
		member.FatherID = req.FatherID
	}
	if req.MotherID != nil {
		member.MotherID = req.MotherID
	}
	if req.Generation != nil {
		member.Generation = *req.Generation
	}
	if req.Position != nil {
		member.Position = *req.Position
	}
	if req.SampleID != nil {
		member.SampleID = req.SampleID
	}

	if err := s.memberRepo.Update(ctx, member); err != nil {
		return nil, errors.WrapDatabaseError(err)
	}

	return s.toMemberResponse(member), nil
}

// DeleteMember deletes a pedigree member
func (s *PedigreeService) DeleteMember(ctx context.Context, memberID string) error {
	memberIDParsed, err := uuid.Parse(memberID)
	if err != nil {
		return errors.NewValidationError("Invalid member ID")
	}

	return s.memberRepo.Delete(ctx, memberIDParsed)
}

// toPedigreeResponse converts a model to DTO
func (s *PedigreeService) toPedigreeResponse(p *model.Pedigree) *dto.PedigreeResponse {
	return &dto.PedigreeResponse{
		ID:          p.ID.String(),
		Name:        p.Name,
		Disease:     p.Disease,
		MemberCount: len(p.Members),
		CreatedAt:   p.CreatedAt,
		UpdatedAt:   p.UpdatedAt,
	}
}

// toPedigreeDetailResponse converts a model to detailed DTO
func (s *PedigreeService) toPedigreeDetailResponse(p *model.Pedigree) *dto.PedigreeDetailResponse {
	members := make([]dto.PedigreeMemberResponse, len(p.Members))
	for i, m := range p.Members {
		members[i] = *s.toMemberResponse(&m)
	}

	return &dto.PedigreeDetailResponse{
		ID:              p.ID.String(),
		Name:            p.Name,
		Disease:         p.Disease,
		Note:            p.Note,
		ProbandMemberID: p.ProbandMemberID,
		Members:         members,
		CreatedAt:       p.CreatedAt,
		UpdatedAt:       p.UpdatedAt,
	}
}

// toMemberResponse converts a member model to DTO
func (s *PedigreeService) toMemberResponse(m *model.PedigreeMember) *dto.PedigreeMemberResponse {
	return &dto.PedigreeMemberResponse{
		ID:             m.ID.String(),
		PedigreeID:     m.PedigreeID.String(),
		SampleID:       m.SampleID,
		Name:           m.Name,
		Gender:         m.Gender,
		BirthYear:      m.BirthYear,
		IsDeceased:     m.IsDeceased,
		DeceasedYear:   m.DeceasedYear,
		Relation:       string(m.Relation),
		AffectedStatus: string(m.AffectedStatus),
		Phenotypes:     m.Phenotypes,
		FatherID:       m.FatherID,
		MotherID:       m.MotherID,
		Generation:     m.Generation,
		Position:       m.Position,
		HasSample:      m.SampleID != nil,
		CreatedAt:      m.CreatedAt,
		UpdatedAt:      m.UpdatedAt,
	}
}

File: apps/web-germline/src/lib/pedigree.ts
import { api } from './api';
import type {
  Pedigree,
  PedigreeDetail,
  PedigreeCreateRequest,
  PedigreeUpdateRequest,
  PedigreeMember,
  PedigreeMemberCreateRequest,
  PedigreeMemberUpdateRequest,
} from '@/types/pedigree';
import type { PaginatedResponse, QueryParams } from '@/types/common';

interface BackendPaginatedData<T> {
  items: T[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
}

interface BackendPedigree {
  id: string;
  name: string;
  disease?: string;
  note?: string;
  member_count?: number;
  proband_member_id?: string;
  created_at: string;
  updated_at: string;
}

interface BackendPedigreeMember {
  id: string;
  pedigree_id: string;
  sample_id?: string;
  name: string;
  gender: string;
  birth_year?: number;
  is_deceased: boolean;
  deceased_year?: number;
  relation: string;
  affected_status: string;
  phenotypes?: Record<string, unknown>;
  father_id?: string;
  mother_id?: string;
  generation: number;
  position: number;
  has_sample: boolean;
  created_at: string;
  updated_at: string;
}

interface BackendPedigreeDetail {
  id: string;
  name: string;
  disease?: string;
  note?: string;
  proband_member_id?: string;
  members: BackendPedigreeMember[];
  created_at: string;
  updated_at: string;
}

function mapMember(m: BackendPedigreeMember): PedigreeMember {
  return {
    id: m.id,
    pedigreeId: m.pedigree_id,
    sampleId: m.sample_id,
    name: m.name,
    gender: m.gender as PedigreeMember['gender'],
    birthYear: m.birth_year,
    isDeceased: m.is_deceased,
    deceasedYear: m.deceased_year,
    relation: m.relation as PedigreeMember['relation'],
    affectedStatus: m.affected_status as PedigreeMember['affectedStatus'],
    phenotypes: m.phenotypes,
    fatherId: m.father_id,
    motherId: m.mother_id,
    generation: m.generation,
    position: m.position,
    hasSample: m.has_sample,
    createdAt: m.created_at,
    updatedAt: m.updated_at,
  };
}

function mapPedigree(p: BackendPedigree): Pedigree {
  return {
    id: p.id,
    name: p.name,
    disease: p.disease,
    memberCount: p.member_count ?? 0,
    createdAt: p.created_at,
    updatedAt: p.updated_at,
  };
}

function mapPedigreeDetail(d: BackendPedigreeDetail): PedigreeDetail {
  return {
    id: d.id,
    name: d.name,
    disease: d.disease,
    note: d.note,
    probandMemberId: d.proband_member_id,
    members: d.members.map(mapMember),
    createdAt: d.created_at,
    updatedAt: d.updated_at,
  };
}

function mapPaginated<T, U>(data: BackendPaginatedData<T>, mapper: (item: T) => U): PaginatedResponse<U> {
  return {
    items: data.items.map(mapper),
    total: data.total,
    page: data.page,
    pageSize: data.page_size,
    totalPages: data.total_pages,
  };
}

export const pedigreeApi = {
  list: async (params?: QueryParams): Promise<PaginatedResponse<Pedigree>> => {
    const data = await api.get<BackendPaginatedData<BackendPedigree>>('/v1/pedigrees', {
      params: params as Record<string, string>,
    });
    return mapPaginated(data, mapPedigree);
  },

  get: async (id: string): Promise<PedigreeDetail> => {
    const data = await api.get<BackendPedigreeDetail>(`/v1/pedigrees/${id}`);
    return mapPedigreeDetail(data);
  },

  create: async (data: PedigreeCreateRequest): Promise<Pedigree> => {
    const result = await api.post<BackendPedigree>('/v1/pedigrees', data);
    return mapPedigree(result);
  },

  update: async (id: string, data: PedigreeUpdateRequest): Promise<Pedigree> => {
    const result = await api.put<BackendPedigree>(`/v1/pedigrees/${id}`, data);
    return mapPedigree(result);
  },

  delete: (id: string) => api.delete(`/v1/pedigrees/${id}`),

  setProband: async (pedigreeId: string, memberId: string): Promise<PedigreeDetail> => {
    const data = await api.put<BackendPedigreeDetail>(`/v1/pedigrees/${pedigreeId}/proband/${memberId}`, {});
    return mapPedigreeDetail(data);
  },
};

export const pedigreeMemberApi = {
  list: async (pedigreeId: string): Promise<PedigreeMember[]> => {
    const data = await api.get<BackendPedigreeMember[]>(`/v1/pedigrees/${pedigreeId}/members`);
    return data.map(mapMember);
  },

  get: async (pedigreeId: string, memberId: string): Promise<PedigreeMember> => {
    const data = await api.get<BackendPedigreeMember>(`/v1/pedigrees/${pedigreeId}/members/${memberId}`);
    return mapMember(data);
  },

  create: async (pedigreeId: string, data: PedigreeMemberCreateRequest): Promise<PedigreeMember> => {
    const payload = {
      ...data,
      birth_year: data.birthYear,
      is_deceased: data.isDeceased,
      deceased_year: data.deceasedYear,
      affected_status: data.affectedStatus,
      father_id: data.fatherId,
      mother_id: data.motherId,
      sample_id: data.sampleId,
    };
    // Remove camelCase fields
    delete (payload as any).birthYear;
    delete (payload as any).isDeceased;
    delete (payload as any).deceasedYear;
    delete (payload as any).affectedStatus;
    delete (payload as any).fatherId;
    delete (payload as any).motherId;
    delete (payload as any).sampleId;
    const result = await api.post<BackendPedigreeMember>(`/v1/pedigrees/${pedigreeId}/members`, payload);
    return mapMember(result);
  },

  update: async (pedigreeId: string, memberId: string, data: PedigreeMemberUpdateRequest): Promise<PedigreeMember> => {
    const payload: Record<string, unknown> = {};
    if (data.name !== undefined) payload.name = data.name;
    if (data.gender !== undefined) payload.gender = data.gender;
    if (data.birthYear !== undefined) payload.birth_year = data.birthYear;
    if (data.isDeceased !== undefined) payload.is_deceased = data.isDeceased;
    if (data.deceasedYear !== undefined) payload.deceased_year = data.deceasedYear;
    if (data.relation !== undefined) payload.relation = data.relation;
    if (data.affectedStatus !== undefined) payload.affected_status = data.affectedStatus;
    if (data.phenotypes !== undefined) payload.phenotypes = data.phenotypes;
    if (data.fatherId !== undefined) payload.father_id = data.fatherId;
    if (data.motherId !== undefined) payload.mother_id = data.motherId;
    if (data.generation !== undefined) payload.generation = data.generation;
    if (data.position !== undefined) payload.position = data.position;
    if (data.sampleId !== undefined) payload.sample_id = data.sampleId;
    const result = await api.put<BackendPedigreeMember>(`/v1/pedigrees/${pedigreeId}/members/${memberId}`, payload);
    return mapMember(result);
  },

  delete: (pedigreeId: string, memberId: string) =>
    api.delete(`/v1/pedigrees/${pedigreeId}/members/${memberId}`),
};
